src/aapns/connection.py

```diff
@@ -0,0 +1,454 @@
+
+import h2.config
+import h2.connection
+import h2.events
+import h2.exceptions
+import h2.settings
+
+from .errors import Blocked, Closed, FormatError, ResponseTooLarge, StreamReset, Timeout
+
+# Apple limits APN payload (data) to 4KB or 5KB, depending.
+# Request header is not subject to flow control in HTTP/2
+# Data is subject to framing and padding, but those are minor.
+MAX_NOTIFICATION_PAYLOAD_SIZE = 5120
+REQUIRED_FREE_SPACE = 6000
+# OK response is empty
+# Error response is short json, ~30 bytes in size
+MAX_RESPONSE_SIZE = 2 ** 16
+# Inbound connection flow control window
+# It's quite arbitrary, guided by:
+# * concurrent requests limit, server limit being 1000 today
+# * expected response size, see above
+CONNECTION_WINDOW_SIZE = 2 ** 24
+# Connection establishment safety time limits
+CONNECTION_TIMEOUT = 5
+TLS_TIMEOUT = 5
+logger = getLogger(__package__)
+
+
+@dataclass(eq=False)
+class Connection:
+    """Encapsulates a single HTTP/2 connection to the APN server.
+
+    Use `Connection.create(...)` to make connections.
+
+    Example use:
+
+        conn = await Connection.create(...)
+        try:
+            await conn.post(request)
+        finally:
+            await conn.close()
+
+    Connection states:
+    * new (not connected)
+    * starting
+    * active
+    * graceful shutdown (to do: https://github.com/python-hyper/hyper-h2/issues/1181)
+    * closing
+    * closed
+    """
+
+    host: str
+    port: int
+    protocol: h2.connection.H2Connection
+    read_stream: asyncio.StreamReader
+    write_stream: asyncio.StreamWriter
+    should_write: asyncio.Event = field(init=False)
+    channels: Dict[int, Channel] = field(default_factory=dict)
+    reader: asyncio.Task = field(init=False)
+    writer: asyncio.Task = field(init=False)
+    closing: bool = False
+    closed: bool = False
+    outcome: Optional[str] = None
+    max_concurrent_streams: int = 100  # initial per RFC7540#section-6.5.2
+    last_stream_id_got: int = -1
+    last_stream_id_sent: int = -1  # client streams are odd
+
+    @classmethod
+    async def create(cls, origin: str, ssl: Optional[SSLContext] = None):
+        """Connect to `origin` and return a Connection"""
+        url = urlparse(origin)
+        if (
+            url.scheme != "https"
+            or not url.hostname
+            or url.username
+            or url.password
+            or url.path
+            or url.params
+            or url.query
+            or url.fragment
+        ):
+            raise ValueError("Origin must be https://<host>[:<port>]")
+
+        host = url.hostname
+        port = url.port or 443
+
+        ssl_context = ssl if ssl else create_ssl_context()
+        if (
+            OP_NO_TLSv1 not in ssl_context.options
+            or OP_NO_TLSv1_1 not in ssl_context.options
+        ):
+            raise ValueError("SSL Context cannot allow TLS 1.0 or 1.1")
+
+        # https://bugs.python.org/issue40111 validate context h2 alpn
```

Comment:
I think you can just set the alpn protocols here, no need to introspect.

Reply:
Thing is, I don't want to modify the ssl context is user passed theirs. 🤷‍♂️